{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f5bed970",
   "metadata": {},
   "source": [
    "# PLAPT affinities only for HRAS-P01112\n",
    "\n",
    "Protein-Ligands Binding Affinity Prediction Using Pretrained Transformers\n",
    "\n",
    "https://www.biorxiv.org/content/10.1101/2024.02.08.575577v3.full\n",
    "\n",
    "https://github.com/trrt-good/WELP-PLAPT/tree/main\n",
    "\n",
    "Predict the binding affinity of the complex ligand - protein (affinity) using SMILES code of the ligand and the protein sequence of the protein. \n",
    "\n",
    "We download the SMILES formulas (strings) from https://hmdb.ca/downloads# as one single file with the SDF format of all the compounds (hmdb_structures_v5.sdf). SDF = a format for ligands including the coordinates of atoms of ligands. We need only the fidls SMILES from each structure/ligand. This script consider that we have all the fields we need inside the multi-SDF file. If no SMILES, we need to modify the script to calculate for each ligand the SMILES.\n",
    "\n",
    "With the pairs SMILES - protein sequence we can predict with the deep learning model PLAPT the logaritm and the binding affinity. The model is using pre-trained transformers like ProtBERT and ChemBERTa to transform the protein sequence and the SMILEs structure into embeddings that are used for the model.\n",
    "\n",
    "This script is calculating the binding affinities using only one protein and multiple ligands. This can help to run multiple scripts for different target proteins.\n",
    "\n",
    "Thus, these are the steps:\n",
    "\n",
    "- Reading the SMILES and other info for ligands.\n",
    "- Reading the sequence for the protein from FASTA file.\n",
    "- Predict the binding affinities.\n",
    "\n",
    "Due to commas from ligand descriptions, we shall use TAB separated files, not CSV."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b82bd24",
   "metadata": {},
   "source": [
    "## Import the libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7ff4cd80",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "import torch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60e8c901",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1be69e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# info we need for ligands (extracted from multisdf_file)\n",
    "LigandInfo = './ligands_hmdb.tsv'\n",
    "\n",
    "# gene, protein and seq\n",
    "BestPredProts = './InfoBestGenes.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65578def",
   "metadata": {},
   "source": [
    "## Get protein sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "331017c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_BestProts = pd.read_csv(BestPredProts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "c69caeb0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MELWRQCTHWLIQCRVLPPSHRVTWDGAQVCELAQALRDGVLLCQLLNNLLPHAINLREVNLRPQMSQFLCLKNIRTFLSTCCEKFGLKRSELFEAFDLFDVQDFGKVIYTLSALSWTPIAQNRGIMPFPTEEESVGDEDIYSGLSDQIDDTVEEDEDLYDCVENEEAEGDEIYEDLMRSEPVSMPPKMTEYDKRCCCLREIQQTEEKYTDTLGSIQQHFLKPLQRFLKPQDIEIIFINIEDLLRVHTHFLKEMKEALGTPGAANLYQVFIKYKERFLVYGRYCSQVESASKHLDRVAAAREDVQMKLEECSQRANNGRFTLRDLLMVPMQRVLKYHLLLQELVKHTQEAMEKENLRLALDAMRDLAQCVNEVKRDNETLRQITNFQLSIENLDQSLAHYGRPKIDGELKITSVERRSKMDRYAFLLDKALLICKRRGDSYDLKDFVNLHSFQVRDDSSGDRDNKKWSHMFLLIEDQGAQGYELFFKTRELKKKWMEQFEMAISNIYPENATANGHDFQMFSFEETTSCKACQMLLRGTFYQGYRCHRCRASAHKECLGRVPPCGRHGQDFPGTMKKDKLHRRAQDKKRNELGLPKMEVFQEYYGLPPPPGAIGPFLRLNPGDIVELTKAEAEQNWWEGRNTSTNEIGWFPCNRVKPYVHGPPQDLSVHLWYAGPMERAGAESILANRSDGTFLVRQRVKDAAEFAISIKYNVEVKHIKIMTAEGLYRITEKKAFRGLTELVEFYQQNSLKDCFKSLDTTLQFPFKEPEKRTISRPAVGSTKYFGTAKARYDFCARDRSELSLKEGDIIKILNKKGQQGWWRGEIYGRVGWFPANYVEEDYSEYC\n",
      "23\n"
     ]
    }
   ],
   "source": [
    "ProtSeqs = df_BestProts['V3'].tolist()\n",
    "print(ProtSeqs[0])\n",
    "print(len(ProtSeqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "99d4e3bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>V1</th>\n",
       "      <th>V2</th>\n",
       "      <th>V3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VAV1</td>\n",
       "      <td>P15498</td>\n",
       "      <td>VAV_HUMAN Proto-oncogene vav OS=Homo sapiens O...</td>\n",
       "      <td>MELWRQCTHWLIQCRVLPPSHRVTWDGAQVCELAQALRDGVLLCQL...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSC1</td>\n",
       "      <td>Q92574</td>\n",
       "      <td>TSC1_HUMAN Hamartin OS=Homo sapiens OX=9606 GN...</td>\n",
       "      <td>MAQQANVGELLAMLDSPMLGVRDDVTAVFKENLNSDRGPMLVNTLV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TPR</td>\n",
       "      <td>P12270</td>\n",
       "      <td>TPR_HUMAN Nucleoprotein TPR OS=Homo sapiens OX...</td>\n",
       "      <td>MAAVLQQVLERTELNKLPKSVQNKLEKFLADQQSEIDGLKGRHEKF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SMARCA4</td>\n",
       "      <td>P51532</td>\n",
       "      <td>SMCA4_HUMAN Transcription activator BRG1 OS=Ho...</td>\n",
       "      <td>MSTPDPPLGGTPRPGPSPGPGPSPGAMLGPSPGPSPGSAHSMMGPS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SETD2</td>\n",
       "      <td>Q9BYW2</td>\n",
       "      <td>SETD2_HUMAN Histone-lysine N-methyltransferase...</td>\n",
       "      <td>MKQLQPQPPPKMGDFYDPEHPTPEEEENEAKIENVQKTGFIKGPMF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>RB1</td>\n",
       "      <td>P06400</td>\n",
       "      <td>RB_HUMAN Retinoblastoma-associated protein OS=...</td>\n",
       "      <td>MPPKTPRKTAATAAAAAAEPPAPPPPPPPEEDPEQDSGPEDLPLVR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PREX2</td>\n",
       "      <td>Q70Z35</td>\n",
       "      <td>PREX2_HUMAN Phosphatidylinositol 3_4_5-trispho...</td>\n",
       "      <td>MSEDSRGDSRAESAKDLEKQLRLRVCVLSELQKTERDYVGTLEFLV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PPP2R1A</td>\n",
       "      <td>P30153</td>\n",
       "      <td>2AAA_HUMAN Serine/threonine-protein phosphatas...</td>\n",
       "      <td>MAAADGDDSLYPIAVLIDELRNEDVQLRLNSIKKLSTIALALGVER...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NBN</td>\n",
       "      <td>O60934</td>\n",
       "      <td>NBN_HUMAN Nibrin OS=Homo sapiens OX=9606 GN=NB...</td>\n",
       "      <td>MWKLLPAAGPAGGEPYRLLTGVEYVVGRKNCAILIENDQSISRNHA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MUTYH</td>\n",
       "      <td>Q9UIF7</td>\n",
       "      <td>MUTYH_HUMAN Adenine DNA glycosylase OS=Homo sa...</td>\n",
       "      <td>MTPLVSRLSRLWAIMRKPRAAVGSGHRKQAASQEGRQKHAKNNSQA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>MARK3</td>\n",
       "      <td>P27448</td>\n",
       "      <td>MARK3_HUMAN MAP/microtubule affinity-regulatin...</td>\n",
       "      <td>MSTRTPLPTVNERDTENHTSHGDGRQEVTSRTSRSGARCRNSIASC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>JAG1</td>\n",
       "      <td>P78504</td>\n",
       "      <td>JAG1_HUMAN Protein jagged-1 OS=Homo sapiens OX...</td>\n",
       "      <td>MRSPRTRGRSGRPLSLLLALLCALRAKVCGASGQFELEILSMQNVN...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HRAS</td>\n",
       "      <td>P01112</td>\n",
       "      <td>RASH_HUMAN GTPase HRas OS=Homo sapiens OX=9606...</td>\n",
       "      <td>MTEYKLVVVGAGGVGKSALTIQLIQNHFVDEYDPTIEDSYRKQVVI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DNM2</td>\n",
       "      <td>P50570</td>\n",
       "      <td>DYN2_HUMAN Dynamin-2 OS=Homo sapiens OX=9606 G...</td>\n",
       "      <td>MGNRGMEELIPLVNKLQDAFSSIGQSCHLDLPQIAVVGGQSAGKSS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CDKN2C</td>\n",
       "      <td>P42773</td>\n",
       "      <td>CDN2C_HUMAN Cyclin-dependent kinase 4 inhibito...</td>\n",
       "      <td>MAEPWGNELASAAARGDLEQLTSLLQNNVNVNAQNGFGRTALQVMK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>P42771</td>\n",
       "      <td>CDN2A_HUMAN Cyclin-dependent kinase inhibitor ...</td>\n",
       "      <td>MEPAAGSSMEPSADWLATAAARGRVEEVRALLEAGALPNAPNSYGR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>CCNE1</td>\n",
       "      <td>P24864</td>\n",
       "      <td>CCNE1_HUMAN G1/S-specific cyclin-E1 OS=Homo sa...</td>\n",
       "      <td>MPRERRERDAKERDTMKEDGGAEFSARSRKRKANVTVFLQDPDEEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CASP8</td>\n",
       "      <td>Q14790</td>\n",
       "      <td>CASP8_HUMAN Caspase-8 OS=Homo sapiens OX=9606 ...</td>\n",
       "      <td>MDFSRNLYDIGEQLDSEDLASLKFLSLDYIPQRKQEPIKDALMLFQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BUB1B</td>\n",
       "      <td>O60566</td>\n",
       "      <td>BUB1B_HUMAN Mitotic checkpoint serine/threonin...</td>\n",
       "      <td>MAAVKKEGGALSEAMSLEGDEWELSKENVQPLRQGRIMSTLQGALA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>BCL10</td>\n",
       "      <td>O95999</td>\n",
       "      <td>BCL10_HUMAN B-cell lymphoma/leukemia 10 OS=Hom...</td>\n",
       "      <td>MEPTAPSLTEEDLTEVKKDALENLRVYLCEKIIAERHFDHLRAKKI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ATG7</td>\n",
       "      <td>O95352</td>\n",
       "      <td>ATG7_HUMAN Ubiquitin-like modifier-activating ...</td>\n",
       "      <td>MAAATGDPGLSKLQFAPFSSALDVGFWHELTQKKLNEYRLDEAPKD...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>Q8IXJ9</td>\n",
       "      <td>ASXL1_HUMAN Putative Polycomb group protein AS...</td>\n",
       "      <td>MKDKQKKKKERTWAEAARLVLENYSDAPMTPKQILQVIEAEGLKEM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ACVR1</td>\n",
       "      <td>Q04771</td>\n",
       "      <td>ACVR1_HUMAN Activin receptor type-1 OS=Homo sa...</td>\n",
       "      <td>MVDGVMILPVLIMIALPSPSMEDEKPKVNPKLYMCVCEGLSCGNED...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene      V1                                                 V2  \\\n",
       "0      VAV1  P15498  VAV_HUMAN Proto-oncogene vav OS=Homo sapiens O...   \n",
       "1      TSC1  Q92574  TSC1_HUMAN Hamartin OS=Homo sapiens OX=9606 GN...   \n",
       "2       TPR  P12270  TPR_HUMAN Nucleoprotein TPR OS=Homo sapiens OX...   \n",
       "3   SMARCA4  P51532  SMCA4_HUMAN Transcription activator BRG1 OS=Ho...   \n",
       "4     SETD2  Q9BYW2  SETD2_HUMAN Histone-lysine N-methyltransferase...   \n",
       "5       RB1  P06400  RB_HUMAN Retinoblastoma-associated protein OS=...   \n",
       "6     PREX2  Q70Z35  PREX2_HUMAN Phosphatidylinositol 3_4_5-trispho...   \n",
       "7   PPP2R1A  P30153  2AAA_HUMAN Serine/threonine-protein phosphatas...   \n",
       "8       NBN  O60934  NBN_HUMAN Nibrin OS=Homo sapiens OX=9606 GN=NB...   \n",
       "9     MUTYH  Q9UIF7  MUTYH_HUMAN Adenine DNA glycosylase OS=Homo sa...   \n",
       "10    MARK3  P27448  MARK3_HUMAN MAP/microtubule affinity-regulatin...   \n",
       "11     JAG1  P78504  JAG1_HUMAN Protein jagged-1 OS=Homo sapiens OX...   \n",
       "12     HRAS  P01112  RASH_HUMAN GTPase HRas OS=Homo sapiens OX=9606...   \n",
       "13     DNM2  P50570  DYN2_HUMAN Dynamin-2 OS=Homo sapiens OX=9606 G...   \n",
       "14   CDKN2C  P42773  CDN2C_HUMAN Cyclin-dependent kinase 4 inhibito...   \n",
       "15   CDKN2A  P42771  CDN2A_HUMAN Cyclin-dependent kinase inhibitor ...   \n",
       "16    CCNE1  P24864  CCNE1_HUMAN G1/S-specific cyclin-E1 OS=Homo sa...   \n",
       "17    CASP8  Q14790  CASP8_HUMAN Caspase-8 OS=Homo sapiens OX=9606 ...   \n",
       "18    BUB1B  O60566  BUB1B_HUMAN Mitotic checkpoint serine/threonin...   \n",
       "19    BCL10  O95999  BCL10_HUMAN B-cell lymphoma/leukemia 10 OS=Hom...   \n",
       "20     ATG7  O95352  ATG7_HUMAN Ubiquitin-like modifier-activating ...   \n",
       "21    ASXL1  Q8IXJ9  ASXL1_HUMAN Putative Polycomb group protein AS...   \n",
       "22    ACVR1  Q04771  ACVR1_HUMAN Activin receptor type-1 OS=Homo sa...   \n",
       "\n",
       "                                                   V3  \n",
       "0   MELWRQCTHWLIQCRVLPPSHRVTWDGAQVCELAQALRDGVLLCQL...  \n",
       "1   MAQQANVGELLAMLDSPMLGVRDDVTAVFKENLNSDRGPMLVNTLV...  \n",
       "2   MAAVLQQVLERTELNKLPKSVQNKLEKFLADQQSEIDGLKGRHEKF...  \n",
       "3   MSTPDPPLGGTPRPGPSPGPGPSPGAMLGPSPGPSPGSAHSMMGPS...  \n",
       "4   MKQLQPQPPPKMGDFYDPEHPTPEEEENEAKIENVQKTGFIKGPMF...  \n",
       "5   MPPKTPRKTAATAAAAAAEPPAPPPPPPPEEDPEQDSGPEDLPLVR...  \n",
       "6   MSEDSRGDSRAESAKDLEKQLRLRVCVLSELQKTERDYVGTLEFLV...  \n",
       "7   MAAADGDDSLYPIAVLIDELRNEDVQLRLNSIKKLSTIALALGVER...  \n",
       "8   MWKLLPAAGPAGGEPYRLLTGVEYVVGRKNCAILIENDQSISRNHA...  \n",
       "9   MTPLVSRLSRLWAIMRKPRAAVGSGHRKQAASQEGRQKHAKNNSQA...  \n",
       "10  MSTRTPLPTVNERDTENHTSHGDGRQEVTSRTSRSGARCRNSIASC...  \n",
       "11  MRSPRTRGRSGRPLSLLLALLCALRAKVCGASGQFELEILSMQNVN...  \n",
       "12  MTEYKLVVVGAGGVGKSALTIQLIQNHFVDEYDPTIEDSYRKQVVI...  \n",
       "13  MGNRGMEELIPLVNKLQDAFSSIGQSCHLDLPQIAVVGGQSAGKSS...  \n",
       "14  MAEPWGNELASAAARGDLEQLTSLLQNNVNVNAQNGFGRTALQVMK...  \n",
       "15  MEPAAGSSMEPSADWLATAAARGRVEEVRALLEAGALPNAPNSYGR...  \n",
       "16  MPRERRERDAKERDTMKEDGGAEFSARSRKRKANVTVFLQDPDEEM...  \n",
       "17  MDFSRNLYDIGEQLDSEDLASLKFLSLDYIPQRKQEPIKDALMLFQ...  \n",
       "18  MAAVKKEGGALSEAMSLEGDEWELSKENVQPLRQGRIMSTLQGALA...  \n",
       "19  MEPTAPSLTEEDLTEVKKDALENLRVYLCEKIIAERHFDHLRAKKI...  \n",
       "20  MAAATGDPGLSKLQFAPFSSALDVGFWHELTQKKLNEYRLDEAPKD...  \n",
       "21  MKDKQKKKKERTWAEAARLVLENYSDAPMTPKQILQVIEAEGLKEM...  \n",
       "22  MVDGVMILPVLIMIALPSPSMEDEKPKVNPKLYMCVCEGLSCGNED...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_BestProts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7996ad1",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_BestProts = df_BestProts[12] # only HRAS-P01112"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c3da84c",
   "metadata": {},
   "source": [
    "## Get ligand SMILES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "626b90ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the file with ligand SMILES\n",
    "df_ligands = pd.read_csv(LigandInfo, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "427f2766",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get only the list of SMILES\n",
    "smiles = list(df_ligands['SMILES'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5dc5dac4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "217776"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the number of ligands / SMILES\n",
    "len(smiles)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60681749",
   "metadata": {},
   "source": [
    "## Calculate affinity ligand - protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e449cde3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in df_BestProts.iterrows():\n",
    "    \n",
    "    from plapt import Plapt\n",
    "    \n",
    "    xgene  = row['gene']\n",
    "    xprot  = row['V1']\n",
    "    xdescr = row['V2']\n",
    "    xseq   = row['V3']\n",
    "    print(f\"\\n-> {index} = Gene:{row['gene']}, Prot:{row['V1']}, Info:{row['V2']}, Seq:{row['V3']}\")\n",
    "    \n",
    "    sequences = [xseq] * len(smiles)\n",
    "    \n",
    "    # set cuda for the calculations\n",
    "    plapt = Plapt(device=\"cuda\")\n",
    "    \n",
    "    # set a timer\n",
    "    start_time = time.time()\n",
    "\n",
    "    # calculate affinities for all pairs of protein - ligand using 2 list of sequences and smiles\n",
    "    results = plapt.predict_affinity(sequences, smiles)\n",
    "\n",
    "    end_time = time.time()\n",
    "    execution_time = end_time - start_time\n",
    "\n",
    "    print(\"Execution time:\", execution_time, \"seconds\")\n",
    "    print(\"Exec time in hours = \", execution_time/60/60)\n",
    "    \n",
    "    \n",
    "    # get the results as dataframe\n",
    "    data = {\"smiles\": smiles, \"neg_log10_affinity_M\": [d[\"neg_log10_affinity_M\"] for d in results], \"affinity_uM\": [d[\"affinity_uM\"] for d in results]}\n",
    "    df_affinities = pd.DataFrame(data)\n",
    "    \n",
    "    # add ligand info columns to the affinity results\n",
    "    df_affinities['DATABASE_ID']  = list(df_ligands['DATABASE_ID'])\n",
    "    df_affinities['HMDB_ID']  = list(df_ligands['HMDB_ID'])\n",
    "    df_affinities['GENERIC_NAME'] = list(df_ligands['GENERIC_NAME'])\n",
    "    \n",
    "    # sort the results by affinities\n",
    "    df_affinities = df_affinities.sort_values(by='affinity_uM')\n",
    "    \n",
    "    # add a column with the index of the proteins in the list\n",
    "    df_affinities['GeneID'] = xgene\n",
    "    df_affinities['ProtID'] = xprot\n",
    "    df_affinities['FastaDescription'] = xdescr\n",
    "    df_affinities['ProtSequence'] = xseq\n",
    "    \n",
    "    outFile = './results/affinities_hmdb_HRAS-P01112.tsv'\n",
    "    df_affinities.to_csv(outFile, sep='\\t', na_rep='N/A', index=False)\n",
    "    \n",
    "    \n",
    "    del sequences\n",
    "    del results\n",
    "    del data\n",
    "    del df_affinities\n",
    "    del plapt\n",
    "    torch.cuda.empty_cache()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44b4eaf6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
